javalib/src/main/scala/java/util/regex/PatternCompiler.scala

```diff
@@ -0,0 +1,1645 @@
+    val state = new CharacterClassState(asciiCaseInsensitive)
+    import state._
+
+    // State that is only used in this function
+    var conjunction = ""
+
+    val pattern = PatternCompiler.this.pattern // local copy
+    val len = pattern.length()
+
+    pIndex += 1 // skip '['
+
+    /* If there is a leading '^' right after the '[', the whole class is
+     * negated. In a sense, '^' is the operator with the lowest precedence.
+     */
+    val isNegated = pIndex != len && pattern.charAt(pIndex) == '^'
+    if (isNegated)
+      pIndex += 1
+
+    while (pIndex != len) {
+      val dispatchCP = pattern.codePointAt(pIndex)
+
+      // startCodePoint will be -1 if it was a control sequence or a predefined character class
+      val startCodePoint: Int = (dispatchCP: @switch) match {
+        case ']' =>
+          pIndex += 1
+          val conjunct = finishConjunct(isNegated)
+          return (if (conjunction == "") conjunct else s"(?:$conjunction$conjunct)")
+
+        case '&' =>
+          pIndex += 1
+          if (pIndex != len && pattern.charAt(pIndex) == '&') {
+            pIndex += 1
+            val conjunct = finishConjunct(isNegated)
+            conjunction += (if (isNegated) conjunct + "|" else s"(?=$conjunct)")
+            thisConjunct = ""
+            thisSegment = ""
+            -1
+          } else {
+            '&'
+          }
+
+        case '[' =>
+          addAlternative(compileCharacterClass())
+          -1
+
+        case '\\' =>
+          pIndex += 1
+          if (pIndex == len)
+            parseError("Illegal escape sequence")
+          val c2 = pattern.charAt(pIndex)
+          (c2: @switch) match {
+            case 'd' | 'D' | 'h' | 'H' | 's' | 'S' | 'v' | 'V' | 'w' | 'W' | 'p' | 'P' =>
+              val cls = parsePredefinedCharacterClass(c2)
+              cls.kind match {
+                case CompiledCharClass.PosP =>
+                  thisSegment += "\\p{" + cls.data + "}"
+                case CompiledCharClass.NegP =>
+                  thisSegment += "\\P{" + cls.data + "}"
+                case CompiledCharClass.PosClass =>
+                  thisSegment += cls.data
+                case CompiledCharClass.NegClass =>
+                  addAlternative(codePointNotAmong(cls.data))
+              }
+              -1
+
+            case 'Q' =>
+              pIndex += 1
+              val end = pattern.indexOf("\\E", pIndex)
+              if (end < 0)
+                parseError("Unclosed character class")
+              while (pIndex != end) {
+                val codePoint = pattern.codePointAt(pIndex)
+                addSingleCodePoint(codePoint)
+                pIndex += charCount(codePoint)
+              }
+              pIndex += 2 // for the \E
+              -1
+
+            case _ =>
+              parseSingleCodePointEscape()
+          }
+
+        case _ =>
+          pIndex += charCount(dispatchCP)
+          dispatchCP
+      }
+
+      if (startCodePoint != -1) {
```

Comment:
IMO this makes the control flow quite hard to understand. Is there any reasonable way we can put range parsing in a helper and invoke it in case we hit a CP escape or a normal character?